File: Backend/Domain/Entities/Product.cs
namespace Domain.Entities;

public class Product : BaseEntity
{
    public Guid Id { get; set; }
    public string Sku { get; set; } = null!;
    public string Name { get; set; } = null!;
    public string? Description { get; set; }
    public decimal Price { get; set; }
    public decimal? DiscountPrice { get; set; }
    public int Stock { get; set; }
    public Guid BrandId { get; set; }
    public Guid CategoryId { get; set; }
    public List<string> ImageUrls { get; set; } = [];
    public List<string> Tags { get; set; } = [];
    public bool IsFeatured { get; set; }

    // Navigation properties
    public Brand Brand { get; set; } = null!;
    public Category Category { get; set; } = null!;
    public ICollection<Review> Reviews { get; set; } = [];
}

File: Frontend/types/product.ts
import { productRequestSchema } from "@/schemas/productSchema";
import { UUID } from "@/types";
import { z } from "zod";
import { BaseGetListRequest } from "./base";

export type ProductResponse = {
  id: UUID;
  sku: string;
  name: string;
  costPrice: number;
  price: number;
  discountPrice?: number;
  category: string;
  brand: string;
  imageUrls: string[];
  rating: number;
  stock: number;
  createdDate: Date;
};

export type ProductListItemResponse = {
  id: UUID;
  name: string;
  price: number;
  discountPrice?: number;
  stock: number;
  brand: string;
  rating: number;
  thumbnailUrl: string;
};

export type ProductDetailResponse = {
  id: UUID;
  sku: string;
  name: string;
  description: string;
  costPrice: number;
  price: number;
  discountPrice?: number;
  stock: number;
  brand: string;
  category: string;
  tags: string[];
  imageUrls: string[];
  rating: number;
  isFeatured?: boolean;
};

export type ProductUpdateResponse = {
  id: UUID;
  sku: string;
  name: string;
  description: string;
  costPrice: number;
  price: number;
  discountPrice?: number;
  stock: number;
  brandId: UUID;
  categoryId: UUID;
  thumbnailUrl: string;
  tags: string[];
  imageUrls: string[];
  isFeatured?: boolean;
};

export enum ProductSort {
  Default,
  PriceAsc,
  PriceDesc,
  Newest,
  Oldest,
}

export type GetListFilteredProductRequest = {
  minPrice?: number;
  maxPrice?: number;
  categories?: string;
  brands?: string;
  sort?: ProductSort;
} & BaseGetListRequest;

export type ProductRequest = z.infer<typeof productRequestSchema>;
